ui/app/src/components/DatasourceList/DatasourceEditorForm.tsx

```diff
@@ -0,0 +1,245 @@
+// Copyright 2023 The Perses Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { useImmer } from 'use-immer';
+import { Datasource, Display } from '@perses-dev/core';
+import {
+  Box,
+  Button,
+  Divider,
+  FormControl,
+  FormHelperText,
+  Grid,
+  InputLabel,
+  MenuItem,
+  Select,
+  Stack,
+  TextField,
+  Typography,
+} from '@mui/material';
+import { DispatchWithoutAction, useCallback, useMemo, useState } from 'react';
+import { PluginEditor } from '@perses-dev/plugin-system';
+import { DiscardChangesConfirmationDialog } from '@perses-dev/components';
+import { useIsReadonly } from '../../model/config-client';
+
+// TODO: Replace with proper validation library
```

Comment:
cc @Gladorme @celian-garcia  I kind of alluded to this in the sync, but if you guys want to try out a form state / validation library in the app, I think that'd be cool. We were trying to avoid it last time this came up until we could do some further research, but these forms are getting more and more complex

If we find one we like, could also potentially use on dashboards later. As far as validation, this [zod](https://github.com/colinhacks/zod) lib has been on my radar and it integrates with a bunch of form packages: https://github.com/colinhacks/zod#form-integrations

Reply:
Yeah I have no knowledge/opinion on React ecosystem. I would trust you.
This zod seems powerful. Reading the doc an idea of having the same validation on Backend and Frontend came to mind. But maybe I'm overestimating the tool :smile:

Reply:
Because developing in the UI, I've been sometimes facing 400 issues that were not discovered by the UI so if any mechanism could ensure the validations in the backend are the same as the validation in the frontend, that would be :fire:

Reply:
+1 definitely would be great, I've [brought this up](https://matrix.to/#/!XNmgYIAndZOkEvQAbb:matrix.org/$WA_zjNYIr4v0lGOryD9NbF6QFcyWq1b4gEoB6F8ebGM?via=matrix.org&via=herail.net) in #perses-dev too, there would be some friction in keeping the two schemas in sync, but it might be worth it regardless. Orr you / @AntoineThebaud should write a CUE -> Zod converter! 😄  Surprised there isn't one yet, it already has integrations to convert from all of these: https://github.com/colinhacks/zod#x-to-zod

Reply:
Yeah maintaining two schemas or maintaining two piece of validation code, I prefer the first :smile: 

Also, we could imagine that the validation is all part of the backend. I think it would be worth it to consider an approach where the frontend validation consist in making a debounced query to the backend (making the query only if you edit and don´t touch for one sec for example)

=> The first solution with zod would be ok if we only speak about static validation, but for dynamic validation it may be more complex.

The Variable injection validation that I'm actually not doing here in this frontend PR is a very good example of dynamic validation that won't be possible with schemas. In the backend we are doing a dynamic validation checking that the variables mentioned in the dashboard are well existing (as external or local vars). But we're not doing that check in the frontend.